api/src/routes/user.ts

```diff
@@ -49,5 +49,69 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.post(
+    '/account/reset',
+    {
+      schema: {
+        response: {
+          200: Type.Object({}),
+          500: Type.Object({
+            message: Type.Literal(
+              'Oops! Something went wrong. Please try again in a moment or contact support@freecodecamp.org if the error persists.'
+            ),
+            type: Type.Literal('danger')
+          })
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        await fastify.prisma.userToken.deleteMany({
+          where: { userId: req.session.user.id }
+        });
+        await fastify.prisma.user.update({
+          where: { id: req.session.user.id },
+          data: {
+            progressTimestamps: [Date.now()],
+            currentChallengeId: '',
+            isRespWebDesignCert: false,
+            is2018DataVisCert: false,
+            isFrontEndLibsCert: false,
+            isJsAlgoDataStructCert: false,
+            isApisMicroservicesCert: false,
+            isInfosecQaCert: false,
+            isQaCertV7: false,
+            isInfosecCertV7: false,
+            is2018FullStackCert: false,
+            isFrontEndCert: false,
+            isBackEndCert: false,
+            isDataVisCert: false,
+            isFullStackCert: false,
+            isSciCompPyCertV7: false,
+            isDataAnalysisPyCertV7: false,
+            isMachineLearningPyCertV7: false,
+            isRelationalDatabaseCertV8: false,
+            // TODO: check if schema should be updated with below field
+            // isCollegeAlgebraPyCertV8: false
+            completedChallenges: [],
+            savedChallenges: [],
+            partiallyCompletedChallenges: [],
+            needsModeration: false
+          }
+        });
+
+        return {};
```

Comment:
Do we need this data calls to be atomic, meaning that both of them should succeed or non?

if so, we need to use https://www.prisma.io/docs/concepts/components/prisma-client/transactions in prisma.


```suggestion
        const deleteToken = fastify.prisma.userToken.deleteMany({
          where: { userId: req.session.user.id }
        });
        const resetProgress = fastify.prisma.user.update({
          where: { id: req.session.user.id },
          data: {
            progressTimestamps: [Date.now()],
            currentChallengeId: '',
            isRespWebDesignCert: false,
            is2018DataVisCert: false,
            isFrontEndLibsCert: false,
            isJsAlgoDataStructCert: false,
            isApisMicroservicesCert: false,
            isInfosecQaCert: false,
            isQaCertV7: false,
            isInfosecCertV7: false,
            is2018FullStackCert: false,
            isFrontEndCert: false,
            isBackEndCert: false,
            isDataVisCert: false,
            isFullStackCert: false,
            isSciCompPyCertV7: false,
            isDataAnalysisPyCertV7: false,
            isMachineLearningPyCertV7: false,
            isRelationalDatabaseCertV8: false,
            // TODO: check if schema should be updated with below field
            // isCollegeAlgebraPyCertV8: false
            completedChallenges: [],
            savedChallenges: [],
            partiallyCompletedChallenges: [],
            needsModeration: false
          }
        });

        return fastify.prisma.$transaction([deleteToken, resetProgress]);
```

Reply:
We decided to use transactions for updates which would end up in a broken state if the queries fail.

Reply:
Yes broken state, if we delete an account but hold the data, we will need to run a mongo query to delete the actual data, otherwise we can't target the data through an endpoint.

we will have a broken state if we delete user is without deleting user data

Reply:
TD;LR, we can't run the query again if user id is deleted, and the data isn't.

Reply:
Yes, I am misunderstanding it, because it's reseting not deleting data 🤣